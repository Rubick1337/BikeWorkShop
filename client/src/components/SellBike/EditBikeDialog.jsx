import React, { useState, useEffect } from 'react';
import { Dialog, DialogActions, DialogContent, DialogTitle, TextField, Button, FormControlLabel, Checkbox, Select, MenuItem, InputLabel, FormControl } from '@mui/material';

export default function EditBikeDialog({ open, handleClose, bike, handleEditBikeSubmit, categories, types }) {
    const [editedBike, setEditedBike] = useState({
        id: '',
        name: '',
        price: '',
        model: '',
        brand: '',
        id_category_bike: '',
        id_type_bike: '',
        img: null,
        inSell: true
    });

    useEffect(() => {
        if (bike) {
            console.log('Bike updated'+ bike.id);
            setEditedBike({
                id: bike.id,
                name: bike.name || '',
                price: bike.price || '',
                model: bike.model || '',
                brand: bike.brand || '',
                id_category_bike: bike.id_category_bike || '',
                id_type_bike: bike.id_type_bike || '',
                img: bike.img || null,
                inSell: bike.inSell || false
            });
        }
    }, [bike, open]);

    const handleChange = (event) => {
        const { name, value, type, checked } = event.target;
        setEditedBike(prev => ({
            ...prev,
            [name]: type === 'checkbox' ? checked : value
        }));
    };

    const handleFileChange = (event) => {
        setEditedBike(prev => ({
            ...prev,
            img: event.target.files[0]
        }));
    };

    const handleSubmit = () => {
        console.log("Измененный велосипед:", editedBike);
        handleEditBikeSubmit(editedBike);
    };

    if (!bike) return null;

    return (
        <Dialog open={open} onClose={handleClose}>
            <DialogTitle>Редактировать велосипед</DialogTitle>
            <DialogContent>
                <TextField
                    label="Название"
                    name="name"
                    value={editedBike.name}
                    onChange={handleChange}
                    fullWidth
                    margin="normal"
                />
                <TextField
                    label="Цена"
                    name="price"
                    value={editedBike.price}
                    onChange={handleChange}
                    fullWidth
                    margin="normal"
                    type="number"
                />
                <TextField
                    label="Модель"
                    name="model"
                    value={editedBike.model}
                    onChange={handleChange}
                    fullWidth
                    margin="normal"
                />
                <TextField
                    label="Бренд"
                    name="brand"
                    value={editedBike.brand}
                    onChange={handleChange}
                    fullWidth
                    margin="normal"
                />
                <div>
                    <FormControl fullWidth margin="normal">
                        <InputLabel>Категория</InputLabel>
                        <Select
                            name="id_category_bike"
                            value={editedBike.id_category_bike}
                            onChange={handleChange}
                            label="Категория"
                        >
                            {categories.map((category) => (
                                <MenuItem key={category.id} value={category.id}>
                                    {category.name}
                                </MenuItem>
                            ))}
                        </Select>
                    </FormControl>
                </div>
                <div>
                    <FormControl fullWidth margin="normal">
                        <InputLabel>Тип</InputLabel>
                        <Select
                            name="id_type_bike"
                            value={editedBike.id_type_bike}
                            onChange={handleChange}
                            label="Тип"
                        >
                            {types.map((type) => (
                                <MenuItem key={type.id} value={type.id}>
                                    {type.name}
                                </MenuItem>
                            ))}
                        </Select>
                    </FormControl>
                </div>
                <div>
                    <label>Фото</label>
                    <input type="file" onChange={handleFileChange} />
                </div>
                <FormControlLabel
                    control={
                        <Checkbox
                            checked={editedBike.inSell}
                            onChange={handleChange}
                            name="inSell"
                            color="primary"
                        />
                    }
                    label="В продаже"
                />
            </DialogContent>
            <DialogActions>
                <Button onClick={handleClose} color="primary">Закрыть</Button>
                <Button onClick={handleSubmit} color="primary">Сохранить</Button>
            </DialogActions>
        </Dialog>
    );
}
