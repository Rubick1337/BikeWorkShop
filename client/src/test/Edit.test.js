import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import EditBikeDialog from '../components/EditBikeDialog/EditBikeDialog';

describe('EditBikeDialog Component', () => {
    const mockClose = jest.fn();
    const mockSubmit = jest.fn();
    const bike = {
        _id: '1',
        name: 'Тестовый велосипед',
        price: 1000, // Убедитесь, что это число
        model: 'Модель 1',
        brand: 'Бренд 1',
        description: 'Описание велосипеда',
        id_category_bike: 'cat1',
        id_type_bike: 'type1',
        img: null,
        inSell: true
    };
    const categories = [{ _id: 'cat1', name: 'Категория 1' }, { _id: 'cat2', name: 'Категория 2' }];
    const types = [{ _id: 'type1', name: 'Тип 1' }, { _id: 'type2', name: 'Тип 2' }];

    beforeEach(() => {
        jest.clearAllMocks(); // Очищаем моки перед каждым тестом
    });

    it('displays the correct bike data from props', () => {
        render(
            <EditBikeDialog
                open={true}
                handleClose={mockClose}
                bike={bike}
                handleEditBikeSubmit={mockSubmit}
                categories={categories}
                types={types}
            />
        );

        // Проверяем, что данные отображаются правильно
        expect(screen.getByLabelText(/Название/i)).toHaveValue(bike.name);
        expect(screen.getByLabelText(/Цена/i)).toHaveValue(bike.price);
        expect(screen.getByLabelText(/Модель/i)).toHaveValue(bike.model);
        expect(screen.getByLabelText(/Бренд/i)).toHaveValue(bike.brand);
        expect(screen.getByLabelText(/Описание/i)).toHaveValue(bike.description);
    });

    it('calls handleClose when close button is clicked', () => {
        render(
            <EditBikeDialog
                open={true}
                handleClose={mockClose}
                bike={bike}
                handleEditBikeSubmit={mockSubmit}
                categories={categories}
                types={types}
            />
        );

        fireEvent.click(screen.getByText(/Закрыть/i));
        expect(mockClose).toHaveBeenCalledTimes(1);
    });

    it('updates the state when input values change', () => {
        render(
            <EditBikeDialog
                open={true}
                handleClose={mockClose}
                bike={bike}
                handleEditBikeSubmit={mockSubmit}
                categories={categories}
                types={types}
            />
        );

        // Изменяем значение в поле "Название"
        fireEvent.change(screen.getByLabelText(/Название/i), { target: { value: 'Обновленный велосипед' } });
        expect(screen.getByLabelText(/Название/i)).toHaveValue('Обновленный велосипед');

        // Изменяем значение в поле "Цена"
        fireEvent.change(screen.getByLabelText(/Цена/i), { target: { value: 1500 } });
        expect(screen.getByLabelText(/Цена/i)).toHaveValue(1500);
    });

    it('calls handleEditBikeSubmit with updated data when save button is clicked', () => {
        render(
            <EditBikeDialog
                open={true}
                handleClose={mockClose}
                bike={bike}
                handleEditBikeSubmit={mockSubmit}
                categories={categories}
                types={types}
            />
        );

        fireEvent.change(screen.getByLabelText(/Название/i), { target: { value: 'Обновленный велосипед' } });
        fireEvent.click(screen.getByText(/Сохранить/i));

        expect(mockSubmit).toHaveBeenCalledWith(expect.objectContaining({ name: 'Обновленный велосипед' }));
    });
});