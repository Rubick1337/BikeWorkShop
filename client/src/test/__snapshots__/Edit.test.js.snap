// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditBikeDialog Component matches snapshot 1`] = `<DocumentFragment />`;
